### PR TITLE
Added enchantments to the mutations

### LINKED ISSUE #38692: Additional mutations: Cephalopod - Defensive Ink Spray/Blind
# Is your feature request related to a problem? Please describe.
As I was playing CDDA and reading the issues I noticed that there could be more mutations in specific trees (rather having more trees) and that some trees have way more mutations that others. I decided to role-play a cephalopod and I noticed that it could have more unique mutations. I realized that a cephalopod  mutant currently lacks a signature mutation: ink/spray (its not that useful or essential, but I think would really add some character for this mutant).

> Fishermen sometimes call cephalopods "inkfish,"

# Describe the solution you'd like

A mutation that has a special attack: ink spray that blinds a valid target and deal no or a single point of damage to the eyes. 
Ideally I would like that it would only activate while counterattacking/blocking or getting attacked while moving, since the nature of ink spray is **defensive** rather than offensive.

# Describe alternatives you've considered
not implementing it

# Additional context

Right now my goal is to enrich the cephalopod mutation tree and this is my first attempt.
I'm actually up to implement this mutation myself. ( i'll poke around the discord for that.)

The actual reason why I'm submitting a issue because I wish to know, if you think that my goal (enriching a specific mutation tree with more mutations) is in line with the CDDA and will not be rejected after its done.

This mutation could also be shared with slime, be cause why not (slime spray instead of ink).

I'm looking forward for your feedback

# Balance 

* The ink sacs should be located somewhere on front torso section, so it only activates when not wearing anything on torso, while being hit on any body part. 
* Is intended to compliment the shell mutation ( cant wear anything on torso)
* The effect is two turn blind.
* Haven't decided on a cooldown or a chance yet.
* I'd say that mutation should be passive (like an adrenaline rush) otherwise it would be quite cumbersome to fire it on demand

### PR DESCRIPTION
**Pull request by OldFriendEr**
* Added `INK_GLANDS`, mutation
* Added `generic_blinding_spray_1` spell
* Added `MEP_BLIND_SPRAY` enchantments should cast generic_blinding_spray_1 spell on hit
* `generic_blinding_spray_1` spell should blind a target
* check if the mutation is activated or passive
* ~Added  cooldown to the mutation enchantments~
* ~Changed cooldown to `once_in` time duration.~
* Changed cooldown to once_in int, because it was not checked every turn, but once every attack
* ~TODO: add coverage requirements ( in my case mutation requires uncovered chest)~ told to make a separate PR
* Mutation trigger message shows up before the spell (first you trigger, then you damage)
* Added player and NPC messages upon enchantment activation with 2 wildcards
  * wildcard 1 string caster name
  * wildcard 2 string target name
* fixed a bug when sell cast spells always triggered 
* fixed a bug where self cast spells did not show a trigger message

failed to add the cooldown due to nature of enchantment cache rebuilding the last_cast values were wiped clean.

#### Summary

```SUMMARY: Features "Allows mutations to cast spells via enchantments"```

#### Purpose of change

I wished to expand on existing mutations by adding a few, the first one was ink gland blind for cephalopods. People suggested to use enchantments and spells instead of a hard-coded, but they failed to mention that the system is not complete :)

#### Describe the solution
Added a new vector member to the mutation_type class enchantments. A mutation can now have a list of enchantments.

#### Describe alternatives you've considered

I considered hard-coding my mutation in a player on_hit event just like spider hair that blinds and stuns.

#### Testing

Created a new character gave him a mutation through debug and spawned a rat. It didn't work.
I debugged that the enchantments system does work and it casts the spell, shows the message.  

Tested the chance, tested with `"once_in" : 10` and without the value, both worked

#### Additional context

~I also foresee a shortcoming that there are no custom messages when the enchantment spell activates.~

resolves #38692

### CODE REVIEW

File: data/json/mutations/mutations.json [hunk 1]
```diff
@@ -2534,6 +2534,17 @@
...
     "category": [ "CEPHALOPOD" ]
   },
+  {
+    "type": "mutation",
+    "id": "INK_GLANDS",
+    "name": "Ink glands",
+    "points": 1,
+    "visibility": 1,
+    "ugliness": 1,
+    "description": "You noticed a several ink glands on your torso. These will spray defensive ink and could blind an attacker in an emergency. The torso must not be covered.",
```
Comment (matskuman5) [inline data/json/mutations/mutations.json]:
```suggestion
    "description": "Several ink glands have grown onto your torso. They can be used to spray defensive ink and blind an attacker in an emergency, as long as the torso isn't covered.",
```
makes it clear that it has to be manually activated

Comment (OldFriendEr) [reply data/json/mutations/mutations.json]: Thank you for the edit the original description was pretty bad 👍 I actually wished for someone to fix it when I wrote it 😄

**Comment by foulman** [comment]
Nice to see someone finally bringing "spells" into mutations! 
I'm aware that there are issues with both messages and the on hit system for enchantments, Chaosvolt and I have noticed both of these issues lately, and I can suggest workarounds to try as follows. 
I would try making the spell a targeted spell with an aoe of 1 or 2, which should choose you, but have the targets set to be hostile creatures in the spell itself. I think that would "hit" them with an area effect attack, cones require aiming, whereas the targeted spell should center on the player if that's how it works.  As far as the message goes, if you put in a sub-spell, with extra_effects, even if it's just empty with a message field, that should trigger and display a custom message.

**Comment by KorGgenT** [comment]
> I'm aware that there are issues with both messages and the on hit system for enchantments, Chaosvolt and I have noticed both of these issues lately

I don't see an issue on the github of such a thing.

> workarounds

Please no, we're working on the code. we don't want workarounds. We don't need them - we can fix it properly. The hit_me effect not targeting the attacker should by default hit the attacker for example.

File: data/json/mutations/mutation_spells.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    "description": "Blind enemies for a short time with an ink gland discharge.",
+    "effect": "cone_attack",
+    "effect_str": "blind",
+    "affected_body_parts": [ "EYES" ],
+    "valid_targets": [ "hostile" ],
+    "max_level": 5,
+    "min_damage": 0,
+    "max_damage": 1,
+    "damage_increment": 0.0,
+    "min_aoe": 0,
```
Comment (KorGgenT) [inline data/json/mutations/mutation_spells.json]: no need to write the default values, that's redundant. If the damage_increment is 0, then you will never get from the minimum of 0 to the maximum of 1. it is reasonable to post a minimum value of 0 if there is a max value of nonzero though.

File: msvc-full-features/Cataclysm-vcpkg.vcxproj [hunk 3]
```diff
@@ -23,7 +23,7 @@
     <ProjectGuid>{2586F32C-9F3B-4410-88D1-4543C38511D7}</ProjectGuid>
     <Keyword>Win32Proj</Keyword>
     <RootNamespace>Cataclysm</RootNamespace>
-    <WindowsTargetPlatformVersion>10.0.17134.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.17763.0</WindowsTargetPlatformVersion>
```
Comment (KorGgenT) [inline msvc-full-features/Cataclysm-vcpkg.vcxproj]: please revert the changes to this file.

File: msvc-full-features/Cataclysm-vcpkg.vcxproj [hunk 4]
```diff
@@ -23,7 +23,7 @@
     <ProjectGuid>{2586F32C-9F3B-4410-88D1-4543C38511D7}</ProjectGuid>
     <Keyword>Win32Proj</Keyword>
     <RootNamespace>Cataclysm</RootNamespace>
-    <WindowsTargetPlatformVersion>10.0.17134.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.17763.0</WindowsTargetPlatformVersion>
     <VcpkgTriplet Condition="'$(Platform)'=='Win32'">x86-windows</VcpkgTriplet>
```
Comment (davidpwbrown) [inline msvc-full-features/Cataclysm-vcpkg.vcxproj]: unstage your changes to this file.

File: src/character.cpp [hunk 5]
```diff
@@ -7264,6 +7264,25 @@ void Character::recalculate_enchantment_cache()
             }
         }
     }
+
+    // get from traits/ mutations
+    for( const auto &mutMap : my_mutations )
```
Comment (KorGgenT) [inline src/character.cpp]: can you remove the auto here in favor of the class? i think this is `mutation_branch`?

File: src/character.cpp [hunk 6]
```diff
@@ -7264,6 +7264,25 @@ void Character::recalculate_enchantment_cache()
...
+    {
+        const mutation_branch &mut = mutMap.first.obj();
+
+        // TODO: move to is_active check
+        // only add the passive or powered active mutations
+        if( mut.activated && !mutMap.second.powered ) {
```
Comment (KorGgenT) [inline src/character.cpp]: here's a good spot for a follow up pr for an `active` enum to use for mutations

File: src/magic_enchantment.cpp [hunk 7]
```diff
@@ -181,6 +181,31 @@ bool enchantment::is_active( const Character &guy, const item &parent ) const
     return false;
 }
 
+// TODO: add move mutation ref and checks here
+bool enchantment::is_active(const Character &guy) const
```
Comment (KorGgenT) [inline src/magic_enchantment.cpp]: i recommend calling this function in the other overload in order to reduce duplicated code.

File: src/magic_enchantment.cpp [hunk 8]
```diff
@@ -181,6 +181,31 @@ bool enchantment::is_active( const Character &guy, const item &parent ) const
...
+{    
+
+    if (active_conditions.second == condition::ALWAYS) {
+        return true;
+    }
+
+    // has no items so item-conditions will fail instantly
+    if (active_conditions.first == has::HELD ||
```
Comment (KorGgenT) [inline src/magic_enchantment.cpp]: i don't think this if statement is totally necessary in this particular function, especially since a goal here would be encapsulating the functions a bit for reuse. We can make another overload (and leave this one) that is `bool enchantment::is_active( const Character &guy, const mut_id &mid ) const` with this little bit in it.

File: data/json/mutations/mutation_spells.json [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    "max_aoe": 0,
+    "aoe_increment": 0.2,
+    "min_range": 0,
+    "max_range": 3,
+    "range_increment": 0.0,
```
Comment (KorGgenT) [inline data/json/mutations/mutation_spells.json]: with a range increment of 0, your range will always be 0 (which means it always targets the caster) even with RANDOM_TARGET because that goes up to your max range in a random directoin.

**Comment by KorGgenT** [comment]
So i'm not going to hold you to this here as it's minor, but as a general style note, the codebase prefers snake_case instead of the normal C++ conventions.

**Comment by KorGgenT** [comment]
oh also you need to astyle

**Comment by foulman** [comment]
> 
> 
> > I'm aware that there are issues with both messages and the on hit system for enchantments, Chaosvolt and I have noticed both of these issues lately
> 
> I don't see an issue on the github of such a thing.
> 
> > workarounds
> 
> Please no, we're working on the code. we don't want workarounds. We don't need them - we can fix it properly. The hit_me effect not targeting the attacker should by default hit the attacker for example.

It was assumed, I believe, that the observed behavior was due to the incomplete implementation and/or that being the intended behavior respectively, not actual bugs, especially since we're probably the only two people that would even touch those aspects of the game.

File: msvc-full-features/Cataclysm-vcpkg.vcxproj (hunk 3 above)
Comment (OldFriendEr) [reply msvc-full-features/Cataclysm-vcpkg.vcxproj]: i'll keep this in mind, thank you

File: src/character.cpp (hunk 5 above)
Comment (OldFriendEr) [reply src/character.cpp]: actually its a element of a map `<mutation_id, mutation_data>`, i'll look into c++ syntax how to declare that, i'll keep in mind that you avoid autos.
I'm planing to work on this on Friday again

**Comment by OldFriendEr** [comment]
Thanks again for the review and tips 🙇. 
I'll return to work on this PR on Friday and during the weekend. Looking forward for it, had a blast yesterday 😃

File: src/character.cpp (hunk 5 above)
Comment (davidpwbrown) [reply src/character.cpp]: itd be a std::pair<mutation_id, mutation_data>

File: src/magic_enchantment.cpp (hunk 7 above)
Comment (OldFriendEr) [reply src/magic_enchantment.cpp]: did as you recommended

File: src/magic_enchantment.cpp (hunk 8 above)
Comment (OldFriendEr) [reply src/magic_enchantment.cpp]: made changes accordingly

File: data/json/mutations/mutation_spells.json (hunk 9 above)
Comment (OldFriendEr) [reply data/json/mutations/mutation_spells.json]: I finally figured it out, if i set min_range 0 my range will be 0 at lvl 0 :)

**Comment by OldFriendEr** [comment]
With the third commit the mutation enchantment system does work, now I need to add a cooldown and a notification message.

File: src/character.cpp [hunk 10]
```diff
@@ -7573,13 +7590,13 @@ int Character::get_armor_fire( body_part bp ) const
...
-    enchantment_cache.cast_hit_you( *this, target.pos() );
+    enchantment_cache.cast_hit_you( *this, target );
 }
 
-void Character::on_hit( Creature * /*source*/, body_part /*bp_hit*/,
+void Character::on_hit( Creature *source, body_part /*bp_hit*/,
                         float /*difficulty*/, dealt_projectile_attack const *const /*proj*/ )
 {
-    enchantment_cache.cast_hit_me( *this );
+    enchantment_cache.cast_hit_me( *this, *source);
```
Comment (BevapDin) [inline src/character.cpp]: You should check that `source` is actually non-null before dereferencing it.

Comment (OldFriendEr) [reply src/character.cpp]: hmm, some of the spells ignores the source, and are cast on self, not sure how to proceed

Comment (BevapDin) [reply src/character.cpp]: Doing `*source` causes undefined behaviour if `source` is a null pointer. You need at least something like `if( source == nullptr ) { return; }`.

Comment (OldFriendEr) [reply src/character.cpp]: the think is i don't want to return, i would  gladly do that inside the cast_hit_me,
can i just check for nullptr here and pass null if its nullptr?
or do i have to create an overload with a single parameter for that?

Comment (OldFriendEr) [reply src/character.cpp]: what i really want to know: will it still be undefined behaviour if i check `if(&source != nullptr)` inside the function?

Comment (anothersimulacrum) [reply src/character.cpp]: Yes, dereferencing a null pointer is undefined behaviour, so you've already triggered it by the time you go into the function.

Comment (OldFriendEr) [reply src/character.cpp]: what are my options? i'm c# dev by trade so I'm not sure how do deal with this situation. Is an overload function the only way?

Comment (KorGgenT) [reply src/character.cpp]: you could pass a pointer as a parameter and then check if it's null inside, or you can write an overload function.

Comment (anothersimulacrum) [reply src/character.cpp]: All you need is this.
```suggestion
    if( source == nullptr ) {
        return;
    }
    enchantment_cache.cast_hit_me( *this, *source);
```

Comment (OldFriendEr) [reply src/character.cpp]: I think i handled this by passing a pointer as @KorGgenT suggested by passing a pointer as a param

File: data/json/mutations/mutations.json [hunk 11]
```diff
@@ -2534,6 +2534,17 @@
...
     "category": [ "CEPHALOPOD" ]
   },
+  {
+    "type": "mutation",
+    "id": "INK_GLANDS",
+    "name": "Ink glands",
+    "points": 1,
+    "visibility": 1,
+    "ugliness": 1,
+    "description": "Several ink glands have grown onto your torso. They can be used to spray defensive ink and blind an attacker in an emergency, as long as the torso isn't covered.",
```
Comment (KorGgenT) [inline data/json/mutations/mutations.json]:
```suggestion
    "description": "Several ink glands have grown onto your torso.  They can be used to spray defensive ink and blind an attacker in an emergency, as long as the torso isn't covered.",
```

File: msvc-full-features/Cataclysm-vcpkg.vcxproj [hunk 12]
```diff
@@ -128,4 +128,4 @@
   </ItemGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
   <ImportGroup Label="ExtensionTargets" />
-</Project>
+</Project>
```
Comment (KorGgenT) [inline msvc-full-features/Cataclysm-vcpkg.vcxproj]:
```suggestion
</Project>

```

Comment (KorGgenT) [inline msvc-full-features/Cataclysm-vcpkg.vcxproj]: this is still different for some reason

**Comment by KorGgenT** [comment]
Just made a slight change to add another space after a period. I've loaded up the game and the ink glands work perfectly, and the code looks good. Other than the small weirdness with the vcxproj file, i'd say it looks good.

File: src/mutation_data.cpp [hunk 13]
```diff
@@ -401,6 +401,7 @@ void mutation_branch::load( const JsonObject &jo, const std::string & )
     optional( jo, was_loaded, "leads_to", additions, trait_reader{} );
     optional( jo, was_loaded, "flags", flags, string_reader{} );
     optional( jo, was_loaded, "types", types, string_reader{} );
+    optional( jo, was_loaded, "enchantments", enchantments);
```
Comment (AMurkin) [inline src/mutation_data.cpp]:
```suggestion
    optional( jo, was_loaded, "enchantments", enchantments );
```

File: src/magic.cpp [hunk 14]
```diff
@@ -1939,6 +1944,11 @@ void fake_spell::load( const JsonObject &jo )
     mandatory( jo, false, "id", temp_id );
     id = spell_id( temp_id );
     optional( jo, false, "hit_self", self, false );
+
+    optional( jo, false, "once_in", trigger_once_in);
```
Comment (AMurkin) [inline src/magic.cpp]:
```suggestion
    optional( jo, false, "once_in", trigger_once_in );
```

File: src/player.cpp [hunk 15]
```diff
@@ -5594,13 +5594,16 @@ void player::on_mutation_gain( const trait_id &mid )
...
     update_type_of_scent( mid );
+    recalculate_enchantment_cache(); // mutations can have enchantments
 }
 
 void player::on_mutation_loss( const trait_id &mid )
 {
     morale->on_mutation_loss( mid );
     magic.on_mutation_loss( mid );
     update_type_of_scent( mid, false );
+
```
Comment (AMurkin) [inline src/player.cpp]:
```suggestion

```

**Comment by OldFriendEr** [comment]
Thanks for the fixes and CR, as for now i'm finished with this PR

File: src/magic.cpp [hunk 16]
```diff
@@ -1939,6 +1944,11 @@ void fake_spell::load( const JsonObject &jo )
     mandatory( jo, false, "id", temp_id );
     id = spell_id( temp_id );
     optional( jo, false, "hit_self", self, false );
+
+    optional( jo, false, "once_in", trigger_once_in );
+    optional( jo, false, "message", trigger_message );
+    optional( jo, false, "npc_message", npc_trigger_message );
+
```
Comment (AMurkin) [inline src/magic.cpp]: one more
```suggestion

```

File: src/magic.cpp [hunk 17]
```diff
@@ -1939,6 +1944,10 @@ void fake_spell::load( const JsonObject &jo )
     mandatory( jo, false, "id", temp_id );
     id = spell_id( temp_id );
     optional( jo, false, "hit_self", self, false );
+
+    optional( jo, false, "once_in", trigger_once_in );
```
Comment (KorGgenT) [inline src/magic.cpp:1948]:
```suggestion
    optional( jo, false, "once_in", trigger_once_in, 1 );
```

File: src/magic.h [hunk 18]
```diff
@@ -96,6 +96,12 @@ struct fake_spell {
     int level = 0;
     // target tripoint is source (true) or target (false)
     bool self = false;
+    // a chance to trigger the enchantment spells
+    int trigger_once_in;
```
Comment (KorGgenT) [inline src/magic.h:100]:
```suggestion
    int trigger_once_in = 1;
```
one in one is just 100%, so this is a reasonable default

File: data/json/mutations/mutation_enchantments.json [hunk 19]
```diff
@@ -0,0 +1,15 @@
+[
+  {
+    "type": "enchantment",
+    "id": "MEP_INK_GLAND_SPRAY",
+    "hit_me_effect": [
+      {
+        "id": "generic_blinding_spray_1",
+        "hit_self": false,
+        "once_in": 15,
+        "message": "Your ink glands spray some ink to %2$s's eyes.",
```
Comment (KorGgenT) [inline data/json/mutations/mutation_enchantments.json:10]:
```suggestion
        "message": "Your ink glands spray some ink into %2$s's eyes.",
```

File: data/json/mutations/mutation_enchantments.json [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+        "npc_message": "%1$s's ink glands spay some ink to %2$s's eyes."
```
Comment (KorGgenT) [inline data/json/mutations/mutation_enchantments.json:11]:
```suggestion
        "npc_message": "%1$s's ink glands spay some ink into %2$s's eyes."
```
